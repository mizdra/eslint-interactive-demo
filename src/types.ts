export type FormData = {
  query: string;
};

export type CursorPosition = {
  styleTop: number;
  styleLeft: number;
};

/**
 * `presetIcons` オプションの要素の型。
 * 要素の型には以下が利用できる。
 * - `string`
 *   - ページのパス (例: `'page'`, `'/other-project/page'`) を表す
 * - `string[]`
 *   - ページのパスの配列 (例: `['page1', 'page2']`)
 * - `() => PresetIconsItem[]`
 *   - 動的に `PresetIconsItem[]` を生成して返す関数
 * - `() => Promise<PresetIconsItem[]>`
 *   - `() => PresetIconsItem[]` の非同期版
 * */
export type PresetIconsItem = string | string[] | (() => PresetIconsItem[]) | (() => Promise<PresetIconsItem[]>);
