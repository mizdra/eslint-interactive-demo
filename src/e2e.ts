import { registerIconSuggestion } from './';
void registerIconSuggestion();
