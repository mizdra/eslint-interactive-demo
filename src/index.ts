export { registerIconSuggestion } from './lib/register';
export { Icon } from './lib/icon';
export { getMemberIcons } from './lib/preset-icon';
