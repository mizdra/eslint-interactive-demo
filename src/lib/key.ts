/** IME による変換中かどうか */
export function isComposing(e: KeyboardEvent): boolean {
  return e.isComposing || (e.key === 'Enter' && e.which === 229);
}
