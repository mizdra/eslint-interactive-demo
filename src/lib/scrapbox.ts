import { CursorPosition } from '../types';
import { Icon, iconLinkElementToIcon } from './icon';

export function scanIconsFromEditor(projectName: string, editor: HTMLElement): Icon[] {
  const iconLinkElements = Array.from(editor.querySelectorAll<HTMLAnchorElement>('a.link.icon'));
  return iconLinkElements.map((iconLinkElement) => iconLinkElementToIcon(projectName, iconLinkElement));
}

export function calcCursorPosition(cursor: HTMLElement): CursorPosition {
  const top = +cursor.style.top.slice(0, -2);
  const left = +cursor.style.left.slice(0, -2);
  return {
    styleTop: top,
    styleLeft: left,
  };
}

// https://scrapbox.io/customize/scrapbox-insert-text よりコピペ。
// Thanks @takker99!
export function insertText(textInput: HTMLTextAreaElement, text: string) {
  textInput.focus();
  textInput.value = text;
  const uiEvent = document.createEvent('UIEvent');
  uiEvent.initEvent('input', true, false);
  textInput.dispatchEvent(uiEvent);
}

export function getEditor(): HTMLElement {
  const editor = document.querySelector<HTMLElement>('.editor');
  if (editor) return editor;
  throw new Error('.editor が存在しません');
}
