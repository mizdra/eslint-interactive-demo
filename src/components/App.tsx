import { FunctionComponent } from 'preact';
import { useCallback, useState } from 'preact/hooks';
import { forwardPartialFuzzyMatcher } from '..';
import { useDocumentEventListener } from '../hooks/useDocumentEventListener';
import { useScrapbox } from '../hooks/useScrapbox';
import { uniqueIcons } from '../lib/collection';
import { Icon } from '../lib/icon';
import { isComposing } from '../lib/key';
import { calcCursorPosition, insertText, scanEmbeddedIcons } from '../lib/scrapbox';
import { CursorPosition, Matcher } from '../types';
import { SearchablePopupMenu } from './SearchablePopupMenu';

const DEFAULT_IS_SUGGESTION_OPEN_KEY_DOWN = (e: KeyboardEvent) => {
  return e.key === 'l' && e.ctrlKey && !e.shiftKey && !e.altKey && !e.metaKey;
};

const DEFAULT_IS_INSERT_QUERY_KEY_DOWN = (e: KeyboardEvent) => {
  if (e.key === 'Enter' && !e.ctrlKey && !e.shiftKey && e.altKey && !e.metaKey) return true;
  if (e.key === 'Enter' && !e.ctrlKey && !e.shiftKey && !e.altKey && e.metaKey) return true;
  return false;
};

export type AppProps = {
  isSuggestionOpenKeyDown?: (e: KeyboardEvent) => boolean;
  isSuggestionCloseKeyDown?: (e: KeyboardEvent) => boolean;
  isInsertQueryKeyDown?: (e: KeyboardEvent) => boolean;
  presetIcons?: Icon[];
  defaultSuggestPresetIcons?: boolean;
  matcher?: Matcher;
};

export const App: FunctionComponent<AppProps> = ({
  isSuggestionOpenKeyDown = DEFAULT_IS_SUGGESTION_OPEN_KEY_DOWN,
  isSuggestionCloseKeyDown,
  isInsertQueryKeyDown = DEFAULT_IS_INSERT_QUERY_KEY_DOWN,
  presetIcons = [],
  defaultSuggestPresetIcons = true,
  matcher = forwardPartialFuzzyMatcher,
}) => {
  const { textInput, cursor, editor, layout, projectName } = useScrapbox();

  const [open, setOpen] = useState(false);
  const [cursorPosition, setCursorPosition] = useState<CursorPosition>({ styleTop: 0, styleLeft: 0 });
  const [embeddedIcons, setEmbeddedIcons] = useState<Icon[]>([]);
  const [suggestPresetIcons, setSuggestPresetIcons] = useState(defaultSuggestPresetIcons);
  const composedMatcher = useCallback(
    (query: string) => {
      const composedIcons = uniqueIcons(suggestPresetIcons ? [...embeddedIcons, ...presetIcons] : embeddedIcons);
      return matcher({ query, composedIcons, presetIcons, embeddedIcons });
    },
    [embeddedIcons, matcher, presetIcons, suggestPresetIcons],
  );
  const [query, setQuery] = useState('');

  const handleSelect = useCallback(
    (icon: Icon) => {
      setOpen(false);
      insertText(textInput, icon.getNotation(projectName));
    },
    [projectName, textInput],
  );

  const handleClose = useCallback(() => {
    setOpen(false);
    textInput.focus();
  }, [textInput]);

  const handleSuggestionOpenKeyDown = useCallback(
    (e: KeyboardEvent) => {
      if (layout !== 'page') return; // エディタのあるページ以外ではキー入力を無視する
      e.preventDefault();
      e.stopPropagation();

      if (!open) {
        // ポップアップが閉じていたら開く
        setCursorPosition(calcCursorPosition(cursor));

        // NOTE: ある行にフォーカスがあると、行全体がテキスト化されてしまい、`scanEmbeddedIcons` で
        // アイコンを取得することができなくなってしまう。そのため、予めフォーカスを外し、フォーカスのあった
        // 行のアイコン記法が画像化されるようにしておく。
        textInput.blur();
        // 画像化されたらエディタを走査してアイコンを収集
        const newEmbeddedIcons = scanEmbeddedIcons(projectName, editor);

        setEmbeddedIcons(newEmbeddedIcons);
        setOpen(true);
        setSuggestPresetIcons(defaultSuggestPresetIcons);
      } else {
        // ポップアップが開いていたら、preset icon の表示・非表示をトグルする
        setSuggestPresetIcons((suggestPresetIcons) => !suggestPresetIcons);
      }
    },
    [cursor, defaultSuggestPresetIcons, editor, open, layout, projectName, textInput],
  );

  const handleInsertQueryKeyDown = useCallback(
    (e: KeyboardEvent) => {
      if (layout !== 'page') return; // エディタのあるページ以外ではキー入力を無視する
      if (!open) return; // ポップアップが閉じていたら無視する
      e.preventDefault();
      e.stopPropagation();
      setOpen(false);
      insertText(textInput, `[${query}.icon]`);
    },
    [layout, open, textInput, query],
  );

  const handleKeydown = useCallback(
    (e: KeyboardEvent) => {
      if (isComposing(e)) return; // IMEによる変換中は何もしない
      if (isSuggestionOpenKeyDown(e)) {
        handleSuggestionOpenKeyDown(e);
      } else if (isInsertQueryKeyDown(e)) {
        handleInsertQueryKeyDown(e);
      }
    },
    [isSuggestionOpenKeyDown, isInsertQueryKeyDown, handleSuggestionOpenKeyDown, handleInsertQueryKeyDown],
  );
  useDocumentEventListener('keydown', handleKeydown, { capture: true });

  return (
    <SearchablePopupMenu
      open={open}
      emptyMessage="キーワードにマッチするアイコンがありません"
      cursorPosition={cursorPosition}
      matcher={composedMatcher}
      onSelect={handleSelect}
      onClose={handleClose}
      onInputQuery={setQuery}
      isSuggestionCloseKeyDown={isSuggestionCloseKeyDown}
    />
  );
};
