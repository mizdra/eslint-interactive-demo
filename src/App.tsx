import { FunctionComponent } from 'preact';
import { useCallback, useMemo, useState } from 'preact/hooks';
import { SuggestionBox, Item } from './components/SuggestionBox';
import { useDocumentEventListener } from './hooks/useDocumentEventListener';
import { uniqBy } from './lib/collection';
import { calcCursorPosition, insertText, scanIconsFromEditor } from './lib/scrapbox';
import { CursorPosition, Icon } from './types';

const cursor = document.querySelector<HTMLElement>('.cursor')!;

function uniqueIcons(icons: Icon[]): Icon[] {
  return uniqBy(icons, (icon) => icon.pagePath);
}

function generateItem(icon: Icon): Item {
  return {
    element: (
      <span>
        <img
          alt={icon.imgAlt}
          title={icon.imgTitle}
          style="width: 1.3em; height: 1.3em; object-fit: contain;"
          src={icon.imgSrc}
        />
        {' ' + icon.pagePath}
      </span>
    ),
    searchableText: icon.pagePath,
  };
}

type AppProps = {
  isSuggestionOpenKeyDown: (e: KeyboardEvent) => boolean;
  presetIcons: Icon[];
  editor: HTMLElement;
  textInput: HTMLTextAreaElement;
};

export const App: FunctionComponent<AppProps> = ({ isSuggestionOpenKeyDown, presetIcons, editor, textInput }) => {
  const [open, setOpen] = useState(false);
  const [cursorPosition, setCursorPosition] = useState<CursorPosition>({ styleTop: 0, styleLeft: 0 });
  const [suggestibleIcons, setSuggestibleIcons] = useState<Icon[]>([]);
  const suggestBoxItems = useMemo(() => suggestibleIcons.map(generateItem), [suggestibleIcons]);
  const [presetAppended, setPresetAppended] = useState(false);

  const handleSelect = useCallback(
    (_item: Item, index: number) => {
      setOpen(false);
      insertText(textInput, suggestibleIcons[index].notation);
    },
    [suggestibleIcons, textInput],
  );

  const handleSelectNonexistent = useCallback(
    (query: string) => {
      setOpen(false);
      insertText(textInput, `[${query}.icon]`);
    },
    [textInput],
  );

  const handleClose = useCallback(() => {
    setOpen(false);
    textInput.focus();
  }, [textInput]);

  const handleKeydown = useCallback(
    (e: KeyboardEvent) => {
      if (!isSuggestionOpenKeyDown(e)) return;
      e.preventDefault();
      e.stopPropagation();

      if (open && !presetAppended) {
        setSuggestibleIcons(uniqueIcons([...suggestibleIcons, ...presetIcons]));
        setPresetAppended(true);
      } else {
        const icons = scanIconsFromEditor(scrapbox.Project.name, editor);
        setCursorPosition(calcCursorPosition(cursor));

        // NOTE: ある行にフォーカスがあると、行全体がテキスト化されてしまい、`scanIconsFromEditor` で
        // アイコンを取得することができなくなってしまう。そのため、予めフォーカスを外し、フォーカスのあった
        // 行のアイコン記法が画像化されるようにしておく。
        textInput.blur();
        setSuggestibleIcons(uniqueIcons(icons));
        setOpen(true);
        setPresetAppended(false);
      }
    },
    [editor, isSuggestionOpenKeyDown, open, presetAppended, presetIcons, suggestibleIcons, textInput],
  );
  useDocumentEventListener('keydown', handleKeydown, { capture: true });

  return (
    <SuggestionBox
      open={open}
      emptyMessage="キーワードにマッチするアイコンがありません"
      items={suggestBoxItems}
      cursorPosition={cursorPosition}
      onSelect={handleSelect}
      onSelectNonexistent={handleSelectNonexistent}
      onClose={handleClose}
    />
  );
};
