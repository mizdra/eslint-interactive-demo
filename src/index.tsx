import { render } from 'preact';
import { App } from './App';
import { pagePathToIcon } from './lib/icon';
import { getEditor } from './lib/scrapbox';

type Options = {
  isSuggestionOpenKeyDown?: (e: KeyboardEvent) => boolean;
  // `string` is `pagePath`
  presetIcons?: string[];
  editor?: HTMLElement;
  textInput?: HTMLTextAreaElement;
  cursor?: HTMLElement;
  scrapbox?: Scrapbox;
};

export function registerIconSuggestion(options?: Options) {
  // 直接 editor に mount すると scrapbox 側の react renderer と干渉して壊れるので、
  // editor 内に差し込んだ container に mount する
  const container = document.createElement('div');
  const editor = options?.editor ?? getEditor();
  editor.appendChild(container);

  const presetIcons = options?.presetIcons?.map((pagePath) => pagePathToIcon(scrapbox.Project.name, pagePath));

  render(
    <App
      isSuggestionOpenKeyDown={options?.isSuggestionOpenKeyDown}
      presetIcons={presetIcons}
      editor={editor}
      textInput={options?.textInput}
      cursor={options?.cursor}
      scrapbox={options?.scrapbox}
    />,
    container,
  );
}
