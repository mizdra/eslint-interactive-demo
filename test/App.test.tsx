import { act, fireEvent, render } from '@testing-library/preact';
import userEvent from '@testing-library/user-event';
import { App, AppProps } from '../src/App';
import { insertText } from '../src/lib/scrapbox';
import { Icon } from './../src/types';
import { createCursor, createEditor, createScrapboxAPI, createTextInput } from './helpers/html';
import { keydownAEvent, keydownCtrlLEvent, keydownEnterEvent, keydownEscapeEvent } from './helpers/key';

jest.mock('../src/lib/scrapbox', () => {
  return {
    ...jest.requireActual('../src/lib/scrapbox'),
    insertText: jest.fn(),
  };
});

// ダミーの props
const presetIcons: Icon[] = [];
const editor = createEditor();
const textInput = createTextInput();
const cursor = createCursor({ styleLeft: 0, styleTop: 0 });
const scrapbox = createScrapboxAPI();
const props = { presetIcons, editor, textInput, cursor, scrapbox };

beforeEach(() => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (insertText as any).mockReset();
});

describe('App', () => {
  describe('初期状態', () => {
    test('SuggestBox が表示されない', () => {
      const { asFragment, queryByRole, queryByTestId } = render(<App {...props} />);
      expect(queryByTestId('popup-menu')).toBeNull();
      expect(queryByRole('textbox')).toBeNull();
      expect(asFragment()).toMatchSnapshot();
    });
    test('isSuggestionOpenKeyDown が真になるようなキーを押すと SuggestBox が表示される', async () => {
      const isSuggestionOpenKeyDown = (e: KeyboardEvent) => e.key === 'a';
      const { asFragment, queryByRole, queryByTestId } = render(
        <App {...props} isSuggestionOpenKeyDown={isSuggestionOpenKeyDown} />,
      );
      await act(() => {
        fireEvent(document, keydownEscapeEvent);
      });
      expect(queryByTestId('popup-menu')).toBeNull();
      expect(queryByRole('textbox')).toBeNull();
      await act(() => {
        fireEvent(document, keydownAEvent);
      });
      expect(queryByTestId('popup-menu')).toBeInTheDocument();
      expect(queryByRole('textbox')).toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe('SuggestBox が表示されている時', () => {
    async function renderApp(props: AppProps) {
      const renderResult = render(<App {...props} />);
      await act(() => {
        fireEvent(document, keydownCtrlLEvent);
      });
      return renderResult;
    }
    test('Escape 押下で SuggestBox が閉じる', async () => {
      const { queryByTestId } = await renderApp({ ...props });
      expect(queryByTestId('popup-menu')).toBeInTheDocument();
      await act(() => {
        fireEvent(document, keydownEscapeEvent);
      });
      expect(queryByTestId('popup-menu')).not.toBeInTheDocument();
    });
    describe('Enter を押下した時', () => {
      test('アイテムが1つもなければ QueryInput に入力した pagePath のアイコンが挿入される', async () => {
        const { getByTestId } = await renderApp({ ...props, editor: createEditor(), presetIcons: [] });
        const queryInput = getByTestId('query-input');
        userEvent.type(queryInput, 'foo');
        expect(queryInput).toHaveValue('foo');
        await act(() => {
          fireEvent(document, keydownEnterEvent);
        });
        expect(insertText).toBeCalledWith(expect.anything(), '[foo.icon]');
      });
      test.todo('アイテムがあれば QueryInput に入力した pagePath のアイコンが挿入される');
    });
    test.todo('isSuggestionOpenKeyDown が真になるようなキーを押下したら、presetIcons が suggest される');
  });
});
