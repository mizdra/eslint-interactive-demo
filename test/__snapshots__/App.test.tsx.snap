// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App 初期状態 SuggestBox が表示されない 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`App 初期状態 isSuggestionOpenKeyDown が真になるようなキーを押すと SuggestBox が表示される 1`] = `
<DocumentFragment>
  <div>
    <div
      class="popup-menu"
      data-testid="popup-menu"
      style="top: 0px;"
    >
      <div
        class="button-container"
        data-testid="button-container"
        style="left: 0px; transform: translateX(-NaN%); color: rgb(238, 238, 238); font-size: 11px; display: inline-block; padding: 0px 5px; cursor: not-allowed; background-color: rgb(85, 85, 85);"
      >
        キーワードにマッチするアイコンがありません
      </div>
      <div
        class="triangle"
        style="left: 0px; border-top-color: #555;"
      />
    </div>
    <form
      class="form-inline"
    >
      <input
        class="form-control"
        data-testid="query-input"
        default="true"
        style="position: absolute; top: 0px; left: 0px; transform: translateX(-NaN%);"
      />
    </form>
  </div>
</DocumentFragment>
`;
